server/vulnerabilities/nvd/cpe_matching_rule.go

```diff
@@ -0,0 +1,109 @@
+package nvd
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/Masterminds/semver"
+	"github.com/facebookincubator/nvdtools/wfn"
+)
+
+// CPEMatchingRuleSpec allows you to match againts a CPE. Version ranges are supported via SemVer constraints.
+type CPEMatchingRuleSpec struct {
+	Vendor   string // Software vendor.
+	Product  string // Software product name.
+	TargetSW string // Target software, this usually corresponds to the target OS.
+
+	// Specifies a version constraint. See https://pkg.go.dev/github.com/Masterminds/semver@v1.5.0#hdr-Checking_Version_Constraints
+	// for reference.
+	SemVerConstraint string
+}
+
+func (rule CPEMatchingRuleSpec) getCPEMeta() *wfn.Attributes {
+	return &wfn.Attributes{
+		Vendor:   rule.Vendor,
+		Product:  rule.Product,
+		TargetSW: rule.TargetSW,
+	}
+}
+
+// CPEMatchingRule allows you to express a matching rule based on some CPE properties, one or more
+// CVEs and one or more SemVer constraint. This is used to 'fix' false positives resulting from bad
+// data in the NVD dataset itself.
+// For example: https://nvd.nist.gov/vuln/detail/CVE-2017-13797, one of the CPE entries specified is
+// cpe:2.3:a:apple:icloud:*:*:*:*:*:*:*:* which will match with any iCloud installation, but the
+// vulnerability in question only affects iCloud on Windows up to 7.0.x.
+type CPEMatchingRule struct {
+	CPESpecs []CPEMatchingRuleSpec
+	// Set of CVEs that this rule targets
+	CVEs map[string]bool
+}
+
+// CPEMatches returns true if both the provided CPE match the rule.
+func (rule CPEMatchingRule) CPEMatches(cpeMeta *wfn.Attributes) bool {
+	if cpeMeta == nil {
+		return false
+	}
+
+	var matches bool
+	for _, spec := range rule.CPESpecs {
+		// The SemVer constraint is validated at instantiation time, so it should be ok to ignore the error.
+		constraint, _ := semver.NewConstraint(spec.SemVerConstraint)
+
+		ver, err := semver.NewVersion(wfn.StripSlashes(cpeMeta.Version))
+		if err != nil {
+			matches = matches || false
+			continue
+		}
+
+		matches = matches || (cpeMeta.MatchWithoutVersion(spec.getCPEMeta()) && constraint.Check(ver))
```

Comment:
If it `cpeMeta.MatchWithoutVersion(spec.getCPEMeta()) && constraint.Check(ver)` then you can `return true`?

Reply:
Fixed in https://github.com/fleetdm/fleet/pull/12551/commits/fad824a8225435573a745a90860c8231d60c91f1